src/EditorFeatures/Core/Formatting/FormatCommandHandler.FormatSelection.cs

```diff
@@ -40,31 +43,64 @@ private bool TryExecuteCommand(FormatSelectionCommandArgs args, CommandExecution
                 return false;
             }
 
-            using (context.OperationContext.AddScope(allowCancellation: true, EditorFeaturesResources.Formatting_currently_selected_text))
+            // we only support single selection for now
+            var selection = args.TextView.Selection.GetSnapshotSpansOnBuffer(args.SubjectBuffer);
+            if (selection.Count != 1)
             {
-                var buffer = args.SubjectBuffer;
+                return false;
+            }
+
+            var token = _listener.BeginAsyncOperation(nameof(ExecuteCommand));
+            _ = ExecuteCommandAsync(args, document, selection.Single()).CompletesAsyncOperation(token);
 
-                // we only support single selection for now
-                var selection = args.TextView.Selection.GetSnapshotSpansOnBuffer(buffer);
-                if (selection.Count != 1)
-                {
-                    return false;
-                }
+            return true;
+        }
 
-                var formattingSpan = selection[0].Span.ToTextSpan();
+        private async Task ExecuteCommandAsync(FormatSelectionCommandArgs args, Document document, SnapshotSpan selectionSpan)
+        {
+            _threadingContext.ThrowIfNotOnUIThread();
 
-                Format(args.TextView, buffer, document, formattingSpan, context.OperationContext.UserCancellationToken);
```

Comment:
>Format [](http://example.com/codeflow?start=16&length=6)

This is not fully synchronous code. Any benefit to making it async again?

Reply:
Seems like if we push the formatting to background thread it may be delayed by other background operations resulting in worse user experience.

Reply:
It makes sense to implement this for TypeScript, but then we should have entirely separate code paths for C# and TS.